components/scream/src/physics/p3/p3_functions_f90.cpp

```diff
@@ -1457,5 +1483,73 @@ Real cxx_exp(Real input)
 #endif
 }
 
+void cloud_water_conservation_f(Real qc_, Real qcnuc_, Real dt, Real* qcaut_, Real* qcacc_, Real* qccol_,
+  Real* qcheti_, Real* qcshd_, Real* qiberg_, Real* qisub_, Real* qidep_)
+  {
+    using P3F = Functions<Real, HostDevice>;
+
+    typename P3F::Spack qc(qc_);
+    typename P3F::Spack qcnuc(qcnuc_);
+    typename P3F::Spack qcaut(*qcaut_);
+    typename P3F::Spack qcacc(*qcacc_);
+    typename P3F::Spack qccol(*qccol_);
+    typename P3F::Spack qcheti(*qcheti_);
+    typename P3F::Spack qcshd(*qcshd_);
+    typename P3F::Spack qiberg(*qiberg_);
+    typename P3F::Spack qisub(*qisub_);
+    typename P3F::Spack qidep(*qidep_);
+
+    P3F::cloud_water_conservation(qc, qcnuc, dt, qcaut, qcacc, qccol, qcheti, qcshd, qiberg, qisub, qidep);
+    *qcaut_ = qcaut[0];
+    *qcacc_ = qcacc[0];
+    *qccol_ = qccol[0];
+    *qcheti_ = qcheti[0];
+    *qcshd_ = qcshd[0];
+    *qiberg_ = qiberg[0];
+    *qisub_ = qisub[0];
+    *qidep_ = qidep[0];
+  }
+
+void rain_water_conservation_f(Real qr_, Real qcaut_, Real qcacc_, Real qimlt_, Real qcshd_,
+  Real dt, Real* qrevp_, Real* qrcol_, Real* qrheti_)
+  {
+    using P3F = Functions<Real, HostDevice>;
+
+    typename P3F::Spack qr(qr_);
+    typename P3F::Spack qcaut(qcaut_);
+    typename P3F::Spack qcacc(qcacc_);
+    typename P3F::Spack qimlt(qimlt_);
+    typename P3F::Spack qcshd(qcshd_);
+    typename P3F::Spack qrevp(*qrevp_);
+    typename P3F::Spack qrcol(*qrcol_);
+    typename P3F::Spack qrheti(*qrheti_);
+
+    P3F::rain_water_conservation(qr, qcaut, qcacc, qimlt, qcshd, dt, qrevp, qrcol, qrheti);
+    *qrevp_ = qrevp[0];
+    *qrcol_ = qrcol[0];
+    *qrheti_ = qrheti[0];
+  }
+
+void ice_water_conservation_f(Real qitot_, Real qidep_, Real qinuc_, Real qiberg_, Real qrcol_, Real qccol_,
+  Real qrheti_, Real qcheti_, Real dt, Real* qisub_, Real* qimlt_)
+{
+    using P3F = Functions<Real, HostDevice>;
+    typename P3F::Spack qitot(qitot_);
```

Comment:
You could clean this up by declaring multiple spacks all at once, allowing you to avoid having to repeat the cumbersome type typename P3F::Spack over and over.